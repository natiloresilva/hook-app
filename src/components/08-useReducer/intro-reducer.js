
// Tenemos nuestro estado inicial
const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}]

// Definimos nuestro reducer
// El Reducer recibe 2 argumentos, un initialState y una acc贸n que va a modificar ese estado. SIEMPRE retorna un nuevo estado! 
const todoReducer = ( state = initialState, action ) => {

    if(action?.type === 'agregar'){
        return [...state, action.payload]
    }

    return state
}

// Inicializaci贸n de mi lista de to do
let todos = todoReducer()


// Nuevo elemento para agregar a la lista
const newTodo = [{
    id: 2,
    todo: 'Comprar leche',
    done: false
}]


// Lo voy a hacer mediante una acci贸n, en la acci贸n debemos mandarle la propiedad type y el payload(puede tener otro nombre)
const agregarTodoAction = {
    type:'agregar',
    payload: newTodo
}

// Para agregar un nuevo todo a nuestra lista debemos llamar al reducer
todos = todoReducer(todos, agregarTodoAction)

//console.log(todos)