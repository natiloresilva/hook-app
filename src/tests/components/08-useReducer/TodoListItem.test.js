const { shallow } = require("enzyme")
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem'
import { demoTodos } from '../../fixtures/demoTodos'

describe('pruebas en <TodoListItem />', () => {

    const handleDelete = jest.fn()
    const handleToggle = jest.fn()

    const wrapper = shallow(
        <TodoListItem
            todo={demoTodos[0]}
            index={0}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
        />
    )


    test('Debe de mostrarse correctamente el componente ', () => {
        expect(wrapper).toMatchSnapshot()
    })



    test('Debe de llamar la funcion handleDelete ', () => {
        wrapper.find('button').simulate('click')
        expect(handleDelete).toHaveBeenCalledWith(1)
    })



    test('Debe de llamar la funcion handleToggle ', () => {
        wrapper.find('p').simulate('click')
        expect(handleToggle).toHaveBeenCalledWith(1)
    })



    test('Debe de mostrar el texto correctamente ', () => {
        const p = wrapper.find('p')
        expect(p.text().trim()).toBe(`1. ${demoTodos[0].desc}`)
    })

    test('debe de tener la clase complete si el TODO.done = true', () => {
        const todo = demoTodos[0]
        todo.done = true

        const wrapper = shallow(
            <TodoListItem
                todo={todo}
            />
        )

        expect(wrapper.find('p').hasClass('complete')).toBe(true)
    })


})
