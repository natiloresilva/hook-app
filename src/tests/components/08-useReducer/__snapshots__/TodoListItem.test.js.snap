// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoListItem /> Debe de mostrarse correctamente el componente  1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
